tikv/write.go

```diff
@@ -214,114 +177,212 @@ func (w *writeLockWorker) run() {
+	wb.lockBatch.set(key, lock.MarshalBinary())
 }
 
-func (store *MVCCStore) loadLocks() error {
-	fileName := store.dir + "/lock_store"
-	f, err := os.Open(fileName)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return nil
+func (wb *writeBatch) Commit(key []byte, lock *mvcc.MvccLock) {
+	if lock.Op != uint8(kvrpcpb.Op_Lock) {
+		userMeta := mvcc.NewDBUserMeta(wb.startTS, wb.commitTS)
+		wb.dbBatch.set(key, lock.Value, userMeta)
+		if lock.HasOldVer {
+			oldKey := mvcc.EncodeOldKey(key, lock.OldMeta.CommitTS())
+			wb.dbBatch.set(oldKey, lock.OldVal, lock.OldMeta.ToOldUserMeta(wb.commitTS))
 		}
-		return errors.Trace(err)
 	}
-	defer f.Close()
-	reader := bufio.NewReader(f)
-	hdrBuf := make([]byte, 8)
-	hdr := (*lockEntryHdr)(unsafe.Pointer(&hdrBuf[0]))
-	var keyBuf, valBuf []byte
-	for {
-		_, err = reader.Read(hdrBuf)
-		if err == io.EOF {
+	wb.lockBatch.delete(key)
+}
+
+func (wb *writeBatch) Rollback(key []byte, deleteLock bool) {
+	rollbackKey := mvcc.EncodeRollbackKey(nil, key, wb.startTS)
+	wb.lockBatch.rollback(rollbackKey)
+	if deleteLock {
+		wb.lockBatch.delete(key)
+	}
+}
+
+func (writer *dbWriter) NewWriteBatch(startTS, commitTS uint64) mvcc.WriteBatch {
+	if commitTS > 0 {
+		writer.updateLatestTS(commitTS)
+	} else {
+		writer.updateLatestTS(startTS)
+	}
+	return &writeBatch{
+		startTS:  startTS,
+		commitTS: commitTS,
+	}
+}
+
+func (writer *dbWriter) getLatestTS() uint64 {
+	return atomic.LoadUint64(&writer.latestTS)
+}
+
+func (writer *dbWriter) updateLatestTS(ts uint64) {
+	latestTS := writer.getLatestTS()
+	if ts != math.MaxUint64 && ts > latestTS {
+		atomic.CompareAndSwapUint64(&writer.latestTS, latestTS, ts)
+	}
+}
+
+const delRangeBatchSize = 4096
+
+func (writer *dbWriter) DeleteRange(startKey, endKey []byte, latchHandle mvcc.LatchHandle) error {
+	keys := make([][]byte, 0, delRangeBatchSize)
+	oldStartKey := mvcc.EncodeOldKey(startKey, maxSystemTS)
+	oldEndKey := mvcc.EncodeOldKey(endKey, maxSystemTS)
+	txn := writer.bundle.DB.NewTransaction(false)
+	reader := dbreader.NewDBReader(startKey, endKey, txn, atomic.LoadUint64(&writer.safePoint.timestamp))
+	keys = writer.collectRangeKeys(reader.GetIter(), startKey, endKey, keys)
+	keys = writer.collectRangeKeys(reader.GetIter(), oldStartKey, oldEndKey, keys)
+	return writer.deleteKeysInBatch(latchHandle, keys, delRangeBatchSize)
+}
+
+func (writer *dbWriter) collectRangeKeys(it *badger.Iterator, startKey, endKey []byte, keys [][]byte) [][]byte {
+	if len(endKey) == 0 {
+		panic("invalid end key")
+	}
+	for it.Seek(startKey); it.Valid(); it.Next() {
+		item := it.Item()
+		key := item.KeyCopy(nil)
+		if exceedEndKey(key, endKey) {
 			break
 		}
-		if err != nil {
-			return errors.Trace(err)
-		}
-		if cap(keyBuf) < int(hdr.keyLen) {
-			keyBuf = make([]byte, hdr.keyLen)
-		}
-		if cap(valBuf) < int(hdr.valLen) {
-			valBuf = make([]byte, hdr.valLen)
+		keys = append(keys, key)
+		if len(keys) == delRangeBatchSize {
```

Comment:
So the keys just part of the result? Should involve multiple `collectRangeKyes` and `deleteKeysInBatch` at L346?